sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/AvailableNowDataStreamWrapper.scala

```diff
@@ -0,0 +1,85 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.execution.streaming
+
+import org.apache.spark.internal.Logging
+import org.apache.spark.sql.connector.read.streaming.{MicroBatchStream, ReadLimit, SparkDataStream, SupportsAdmissionControl, SupportsTriggerAvailableNow}
+import org.apache.spark.sql.connector.read.streaming
+
+/**
+ * This class wraps a [[SparkDataStream]] and makes it support Trigger.AvailableNow, by
+ * overriding its [[latestOffset]] method to always return the latest offset when the method is
+ * first called. It is as if there is no new data coming in from the source after the first
+ * [[latestOffset]] call.
+ */
+class AvailableNowDataStreamWrapper(val delegate: SparkDataStream)
+  extends SparkDataStream with SupportsTriggerAvailableNow with Logging {
+
+  private var fetchedOffset: Option[streaming.Offset] = _
+
+  override def initialOffset(): streaming.Offset = delegate.initialOffset()
+
+  override def deserializeOffset(json: String): streaming.Offset = delegate.deserializeOffset(json)
+
+  override def commit(end: streaming.Offset): Unit = delegate.commit(end)
+
+  override def stop(): Unit = delegate.stop()
+
+  override def prepareForTriggerAvailableNow(): Unit = {}
```

Comment:
As this implements `SupportsTriggerAvailableNow`, but seems `prepareForTriggerAvailableNow` is useless? Why don't we record latest offset here?

Reply:
Thanks for the suggestion! This made me realized that we might need to have `startOffset` as an argument in `prepareForTriggerAvailableNow`

Reply:
Turned out `startOffset` is not needed in `prepareForTriggerAvailableNow`. See the latest code update.